dhis-2/dhis-support/dhis-support-system/src/main/java/org/hisp/dhis/system/util/ReflectionUtils.java

```diff
@@ -576,4 +576,57 @@ public static Type getInnerType( ParameterizedType parameterizedType )
 
         return innerType.getActualTypeArguments()[0];
     }
+
+    /**
+     * Get type parameters from a Method. This is mostly useful for getting
+     * return types from methods which returns a generic map for example.
+     *
+     * This method does not guarantee the correct order of parameters. It will
+     * take the super classes / interfaces first, then add the types of the
+     * concrete class.
+     *
+     * @param method Method to return parameter types for
+     * @return List of resolved parameters
+     */
+    public static List<Class<?>> getActualTypeArguments( Method method )
```

Comment:
This more works by accident for simple cases. 

Doesn't make much sense from an algorithmic point of view which is my main concern here. Whoever will try to understand the logic must fail because there is no logic where this algorithm does something remotely correct for the java type system. 

So my thought are more towards making a "simplified" implementation that deliberately just supports the most common use case of inheritance from a non generic starting point (generic superclass or interface).

```java
class A<T> {
  List<T> getElements();
}
class B<T> extends A<T> {}

class Bean extends B<String> {

}
```
So now we want to know what `T` is for the `getElements()` method when we have a `Bean` instance.

First we make a list of the `Type` hierarchy, `Bean` => `B<String>` => `A<String>` - the hard part here is finding the `A<String>` since `B` is extending `A` with a type variable as generic. So here we can simplify and either assume if that is used we have in-order substitution so a `B<String>` becomes a `A<String>` or we can use the names which is correct but a bit harder. Also how do we represent the type `A<String>` if we have no actual instance we can  get from somewhere, where it already exists. It should be clear that this requires some record type
```java
class ActualType {
   Type genericType;
   Type[] actualTypeArguments;
}
```
Then we can create such an `ActualType` from the type `A<T>` we get when asking super-classes or super-interfaces. And the `actualTypeArguments` we can pass down once we found them at the root.
Obviously this isn't always a 1:1 but in a simple implementation we could even assume it is.

So given we have our list `Bean` => `B<String>` => `A<String>` we then just as use the declaring class of method to find that we want to go to `A` and again have to match the `T` to the index of the generic declaration of `A` to find the index we want of the actual types and then we can pick the actual type argument. 

That is surely more complicated to implement then the below but then at least there would be a reasonable relation to the java type system.